Index: src/arch/fermi/emu/emu.c
===================================================================
--- src/arch/fermi/emu/emu.c	(revision 1468)
+++ src/arch/fermi/emu/emu.c	(working copy)
@@ -106,9 +106,13 @@
 		frm_grid_free(frm_emu->grid_list_head);
 
 	/* Free CUDA object list */
-	cuda_object_free_all();
-	linked_list_free(cuda_object_list);
+	if(cuda_object_list) {
+		cuda_object_free_all();
+		linked_list_free(cuda_object_list);
+	}
 
+	m2s_timer_free(frm_emu->timer);
+
 	frm_isa_done();
 
         mem_free(frm_emu->const_mem);
